tc-messaging/src/main/java/com/tc/bytes/TCReferenceSupport.java

```diff
@@ -0,0 +1,267 @@
+ *  License. You may obtain a copy of the License at
+ *
+ *  http://terracotta.org/legal/terracotta-public-license.
+ *
+ *  Software distributed under the License is distributed on an "AS IS" basis,
+ *  WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for
+ *  the specific language governing rights and limitations under the License.
+ *
+ *  The Covered Software is Terracotta Core.
+ *
+ *  The Initial Developer of the Covered Software is
+ *  Terracotta, Inc., a Software AG company
+ *
+ */
+package com.tc.bytes;
+
+import com.tc.util.Assert;
+import com.tc.util.concurrent.SetOnceFlag;
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ */
+public class TCReferenceSupport {
+  private final Queue<TCByteBuffer> returns;
+  private final Collection<TCByteBuffer> items;
+  private final AtomicInteger referenceCount = new AtomicInteger();
+  
+  // Ideally the below variables are not needed.  These are here to track references
+  // that are created but not closed, holding resources.
+  Optional<MemoryTracker> track;
+  
+  private static final Logger LOGGER = LoggerFactory.getLogger(TCReferenceSupport.class);
+  private static final Set<TCReferenceSupport> COMMITTED_REFERENCES = ConcurrentHashMap.newKeySet();
+  private static volatile boolean TRACK_REFERENCES;
+
+  private TCReferenceSupport(Collection<TCByteBuffer> tracked, Queue<TCByteBuffer> returns) {
+    this.items = tracked;
+    this.returns = returns;
+    if (TRACK_REFERENCES) {
+      track = Optional.of(new MemoryTracker());
+    } else {
+      track = Optional.empty();
+    }
+  }
+  
+  public static void startMonitoringReferences() {
+    TRACK_REFERENCES = true;
+  }
+  
+  public static void stopMonitoringReferences() {
+    TRACK_REFERENCES = false;
+    COMMITTED_REFERENCES.clear();
+  }
+  
+  public static int checkReferences() {
+    return COMMITTED_REFERENCES.stream().map(TCReferenceSupport::gc).reduce(0, Integer::sum);
+  }
+  /**
+   * General reference counter.  Starts with a reference count of one.  Each duplicate reference 
+   * increases the reference count by one.  Every close decreases by one.  Once all the references
+   * are closed, the byte buffers are recycled back to the return queue.
+   * 
+   * @param tracked
+   * @param returns
+   * @return 
+   */
+  public static TCReference createReference(Collection<TCByteBuffer> tracked, Queue<TCByteBuffer> returns) {
+    if (returns == null) {
+      returns = new LinkedList<>();
```

Comment:
If we make this a `Consumer` then this would become `buffer -> {}` right?

Reply:
will change